Source/Dafny/DafnyAst.cs

```diff
@@ -325,30 +325,67 @@ public class BuiltIns
       return new ArrowType(f.tok, atd, f.Formals.ConvertAll(arg => Resolver.SubstType(arg.Type, typeMap)), Resolver.SubstType(f.ResultType, typeMap));
     }
 
-    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType) {
+    private object MakeTupleKey(List<bool> isGhost, int dims) {
+      if (dims == 0) {
+        return 0;
+      } else {
+        var g = isGhost[dims - 1];
+        return Tuple.Create(g, MakeTupleKey(isGhost, dims - 1));
+      }
+    }
+
+    public TupleTypeDecl TupleType(IToken tok, int dims, bool allowCreationOfNewType, List<bool> ghostArguments = null) {
       Contract.Requires(tok != null);
       Contract.Requires(0 <= dims);
       Contract.Ensures(Contract.Result<TupleTypeDecl>() != null);
 
       TupleTypeDecl tt;
+      if (ghostArguments == null) {
+        ghostArguments = new bool[dims].ToList().ConvertAll(_ => false);
+      }
+      object key = MakeTupleKey(ghostArguments, dims);
       if (!tupleTypeDecls.TryGetValue(dims, out tt)) {
         Contract.Assume(allowCreationOfNewType);  // the parser should ensure that all needed tuple types exist by the time of resolution
-        if (dims == 2) {
+        if (dims == 2 && !ghostArguments.Contains(true)) {
           // tuple#2 is already defined in DafnyRuntime.cs
-          tt = new TupleTypeDecl(dims, SystemModule, DontCompile());
+          tt = new TupleTypeDecl(ghostArguments, SystemModule, DontCompile());
         } else {
-          tt = new TupleTypeDecl(dims, SystemModule, null);
+          tt = new TupleTypeDecl(ghostArguments, SystemModule, null);
         }
         tupleTypeDecls.Add(dims, tt);
         SystemModule.TopLevelDecls.Add(tt);
       }
       return tt;
     }
 
-    public static string TupleTypeName(int dims) {
-      Contract.Requires(0 <= dims);
-      return "_tuple#" + dims;
+    public static char IsGhostToChar(bool isGhost) {
+      return isGhost ? 'G' : 'O';
+    }
+
+    public static bool IsGhostFromChar(char c) {
+      Contract.Requires(c == 'G' || c == 'O');
+      return c == 'G';
+    }
+
+    public static string GhostArgumentsToString(List<bool> ghostArguments) {
+      return ghostArguments.Count + (!ghostArguments.Contains(true)
+        ? "" : String.Concat(ghostArguments.ConvertAll(IsGhostToChar)));
+    }
+
+    public static List<bool> GhostArgumentsFromString(string s, int count) {
```

Comment:
Consider returning an `IEnumerable<string>`GhostArgumentsFromString, that way the caller can decide whether to do the extra work of allocating a new List.

If you look at the call site of this method, you can see the return value is immediately passed to `Zip` so storing the bool in a List was not needed.

```
        List<bool> ghostArguments = BuiltIns.GhostArgumentsFromString(Name, TypeArgs.Count);
        return "(" + Util.Comma(System.Linq.Enumerable.Zip(TypeArgs, ghostArguments), ...
```

Reply:
I changed the types to `IEnumerable<bool>`.